python/lsst/pipe/base/execution_reports.py

```diff
@@ -0,0 +1,438 @@
+# This file is part of pipe_base.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from __future__ import annotations
+
+__all__ = (
+    "QuantumGraphExecutionReport",
+    "TaskExecutionReport",
+    "DatasetTypeExecutionReport",
+    "lookup_quantum_data_id",
+)
+
+import dataclasses
+import itertools
+import logging
+import uuid
+from collections.abc import Iterable, Mapping
+from typing import Any
+
+import networkx
+import yaml
+from lsst.daf.butler import Butler, DataCoordinate, DatasetRef
+from lsst.resources import ResourcePathExpression
+
+from .graph import QuantumGraph, QuantumNode
+from .pipeline import PipelineDatasetTypes
+
+
+@dataclasses.dataclass
+class DatasetTypeExecutionReport:
+    """A report on the number of produced datasets as well as the status of
+    missing datasets based on metadata.
+
+    A DatasetTypeExecutionReport is created for each DatasetType in a
+    TaskExecutionReport.
+
+    See Also
+    --------
+    TaskExecutionReport
+    QuantumGraphExecutionReport
+    """
+
+    missing_failed: set[DatasetRef] = dataclasses.field(default_factory=set)
+    """Datasets not produced because their quanta failed directly in this
+    run (`set`).
+    """
```

Comment:
```suggestion
    """

```
Should have a blank line after the docstring of one attribute and the definition of the next.